### ISSUE TITLE
a) Feature Request: Function sample_dirichlet, b) State of probabilistic model support?

### ISSUE THREAD

**Issue by andife** [onnx/onnx#6302]
I am very interested in converting DeepLearning models, that contain the PowerSpherical function (https://github.com/andife/power_spherical) to ONNX.

Currently this fails because of the Dirichlet function (https://github.com/pytorch/pytorch/issues/116336).
After my research, I came across https://github.com/onnx/onnxmltools/issues/549, among others
and wondered whether it would be useful to have gamma, dirichlet, beta available in General?
For this reason, the question arises as to what the current state of probabilistic model support looks like?

Dirichlet is available in 
pytorch: https://pytorch.org/docs/stable/distributions.html#dirichlet
tensorflow: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/distributions/dirichlet_multinomial.py

Would it be a suitable direction, e.g. to create a sample Dirichlet method as an Onnx function based on RandomUniform (https://onnx.ai/onnx/operators/onnx__RandomUniform.html#l-onnx-doc-randomuniform)?

**Comment by justinchuby** [comment]
cc @gramalingam

**Comment by gramalingam** [comment]
How easy is it to generate a dirichlet distribution from a uniform distribution? If the function-based implementation is a practical way of computing the distribution, it is a reasonable option to take.

**Comment by Craigacp** [comment]
What are the properties of the RandomUniform op? Is it stateful (i.e. the initial seed is used to generate a value which stays in the op and is updated each time the op is executed), or is it stateless (i.e. the seed should change each time to make the sampled value different)? From reading the docs I assume stateful as otherwise the seed should be an input not an attribute, but I can't think of any other ONNX ops which are which is a bit surprising.

Sampling from a dirichlet basically means sampling from a bunch of gamma distributions, which in turn involves sampling from normal distributions and uniform distributions. The code that does this in numpy is in Cython and not particularly easy to read, this is a [Java implementation](https://github.com/oracle/sandwood/blob/main/Sandwood/runtime/src/main/java/org/sandwood/runtime/internal/numericTools/DistributionSampling.java#L368) which should be relatively straightforward to read.

**Comment by xadupre** [comment]
To generate a direchlet law, we need to generate many exponential law to generate gamma wariable Γ(α, θ). When α is not an integer, it requires to reject some values (reject algorithm). So we don't know how many uniform variables we need to generate one dirichlet variable. I suggest creating a new op to generate a gamma law. Then we can create a dirichlet function.

**Comment by cbourjau** [comment]
> What are the properties of the RandomUniform op? Is it stateful (i.e. the initial seed is used to generate a value which stays in the op and is updated each time the op is executed), or is it stateless (i.e. the seed should change each time to make the sampled value different)? From reading the docs I assume stateful as otherwise the seed should be an input not an attribute, but I can't think of any other ONNX ops which are which is a bit surprising.

That is a very good question! I am not aware of the standard specifying this explicitly. Could you maybe create a dedicated issue for how to deal with PRNG states in the standard, @Craigacp ? It may also be relevant that the attributes of `RandomUniform` may become inputs, as discussed [here](https://github.com/onnx/onnx/issues/5034).

**Comment by Craigacp** [comment]
I've created an issue for the spec clarification, see #6408.
